[opt](kerberos) use ticket cache instead of principal+keytab on BE side

morningman opened the pull request:
### What problem does this PR solve?

#### Overview

Previously, BE node use principal and keytab to do the kerberos authentication.
But only the modified hadoop libhdfs support authenticating in this way, the origin libhdfs
only support setting kerberos ticket cache path, or use system level kerberos authentication context.

This pull request introduces a comprehensive Kerberos authentication module for the BE.
The module is designed to handle Kerberos ticket management, including initialization, authentication, and periodic ticket refresh.
It provides a robust interface for integrating Kerberos authentication, ensuring secure and efficient credential management.

#### Key Components

1. **KerberosConfig** (`kerberos_config.h` and `kerberos_config.cpp`):
   - This class encapsulates the configuration settings required for Kerberos authentication, such as principal, keytab path, and refresh intervals.
   - Provides methods to set and retrieve configuration parameters.

2. **KerberosTicketCache** (`kerberos_ticket_cache.h` and `kerberos_ticket_cache.cpp`):
   - Manages the Kerberos ticket cache, including initialization, login, and periodic refresh of tickets.
   - Supports operations like writing to the ticket cache and checking if a refresh is needed.
   - Utilizes a background thread to periodically refresh tickets based on configured intervals.
   - The default cache file will be written in `/tmp` dir, but can be modified using `kerberos_ccache_path` in be.conf

3. **KerberosTicketMgr** (`kerberos_ticket_mgr.h` and `kerberos_ticket_mgr.cpp`):
   - Acts as a manager for multiple Kerberos ticket caches, handling their lifecycle, including creation, access, and cleanup.
   - Provides methods to get or set ticket caches and retrieve cache file paths.
   - Includes a background thread for cleaning up expired ticket caches every 1 hour. If a cache is longer being referenced, it will be removed.

4. **HdfsMgr**
   - A simple and new class to manager the hdfs fs handler.
   - It replace the old `HdfsHandlerCache`
   - It will check HdfsHandler every 1 hour, and remove unused HdfsHandler after 24 hours.

#### Mainly Changes

1. Introduce a comprehensive kerberos ticket cache management on BE side
1. Use ticket cache path instead of principal and keytab to do the kerberos authentication of libhdfs.
2. Fix the issue that `kerberos_krb5_conf_path` in be.conf does not take effect.
3. Add a new system table `backend_kerberos_ticket_cache`, to view the krb ticket cache of each backend:

```
Doris > select * from information_schema.backend_kerberos_ticket_cache\G
*************************** 1. row ***************************
                  BE_ID: 1738304534666
                  BE_IP: 172.20.32.136
              PRINCIPAL: hdfs/master-1-1@EMR.C-0596176698BD4D17.COM
                 KEYTAB: /path/to/hdfs.keytab
      SERVICE_PRINCIPAL: krbtgt/EMR@EMR.C-0596176698BD4D17.COM
      TICKET_CACHE_PATH: /tmp/doris_krb_ce93d5ebb2a6554c7ba9f43aee3a9e6c
              HASH_CODE: ce93d5ebb2a6554c7ba9f43aee3a9e6c
             START_TIME: 2025-02-01 00:08:26
            EXPIRE_TIME: 2025-02-01 00:09:26
              AUTH_TIME: 2025-02-01 00:08:26
              REF_COUNT: 1
REFRESH_INTERVAL_SECOND: 3600
```

#### Usage

The user interface remains unchanged.
1. set krb5.conf path in be.conf `kerberos_krb5_conf_path`, default is `/etc/krb5.conf`
2. provide kerberos principal the keytab path as usual.

#### Configurations

be.conf

1. `kerberos_ccache_path`
	The dir where kerber ticket cache file saved. the file name as format `doris_krb_xxxx`

2. `kerberos_krb5_conf_path`
	The path of krb5.conf file

6. `kerberos_refresh_interval_second`
	The min interval to refresh a kerberos ticket cache file. default is 1h.

7. cleanup logic

	If the ticket cache is not used for 1 day, it will be deleted.

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [x] Regression test
    - [x] Unit Test
    - [x] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [x] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

hello-stephen commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

morningman committed:
[opt](kerberos) use ticket cache instead of keytab on BE side

morningman commented:
run buildall

morningman commented:
run buildall

morningman commented:
run buildall

morningman commented:
This pull request introduces several changes to enhance the Kerberos authentication functionality and configuration. The key changes include adding new configuration options for Kerberos, implementing a new `KerberosConfig` class, and adding a `KerberosTicketCache` class to manage Kerberos ticket caches, including periodic refresh.

### Kerberos Configuration Enhancements:
* Added `kerberos_refresh_interval_second` configuration option to specify the interval for renewing the Kerberos ticket cache (`be/src/common/config.cpp`, `be/src/common/config.h`). [[1]](diffhunk://#diff-b626e6ab16bc72abf40db76bf5094fcc8ca3c37534c2eb83b63b7805e1b601ffL1143-R1145) [[2]](diffhunk://#diff-46e8c1ada0d43acf8c2965e46e90909089aada1f46531976c10605b837f8da3dR1198-R1199)

### New `KerberosConfig` Class:
* Introduced `KerberosConfig` class to manage Kerberos configuration settings, including the principal, keytab path, krb5.conf path, and refresh intervals (`be/src/common/kerberos/kerberos_config.cpp`, `be/src/common/kerberos/kerberos_config.h`). [[1]](diffhunk://#diff-6c02165c0367e6dc3891d9159a9a101122931c20ad83bc15f072bcd12ac7f392R1-R30) [[2]](diffhunk://#diff-ecc41cb5a03e99e19eb1d829f78a4bbeb9974f8379e6ec4695156eea8f3b431eR1-R69)

### New `KerberosTicketCache` Class:
* Implemented `KerberosTicketCache` class to handle Kerberos ticket cache management, including initialization, login, periodic refresh, and ticket cache path setup (`be/src/common/kerberos/kerberos_ticket_cache.cpp`, `be/src/common/kerberos/kerberos_ticket_cache.h`). [[1]](diffhunk://#diff-502a85fe66921ce7ae5d57b4b607c62adfc8433db067c0d8294692c57a245d39R1-R296) [[2]](diffhunk://#diff-66481c06c0a08665aead28519b09a294148f38a2974a9086403489b03e20fa77R1-R139)

### Build Configuration:
* Added `-DUSE_DORIS_HADOOP_HDFS` definition to the build configuration for Hadoop HDFS support (`be/CMakeLists.txt`).

morningman commented:
run buildall

morningman commented:
run buildall

morningman commented:
run buildall

morningman commented:
run buildall

morningman commented:
run buildall

morningman commented:
run buildall

morningman commented:
run buildall

CalvinKirs commented:
run external

CalvinKirs commented:
run buildall

CalvinKirs commented:
run buildall

CalvinKirs commented:
run buildall

CalvinKirs commented:
run external

CalvinKirs commented:
run external

morningman committed:
fix temp kerberos ticket patch string issue

morningman commented:
run buildall

dataroaring reviewed (approved):
LGTM
